src/et_coupling_pod_utils.F

```diff
@@ -0,0 +1,318 @@
+      INTEGER                                            :: n_b_row, n_b_col
+      INTEGER                                            :: n_p_row, n_p_col
+      INTEGER                                            :: p_row_0, p_col_0
+      INTEGER                                            :: p_row_i, p_col_i
+
+      ! Number of rows/columns in Scalapack blocks
+      n_b_row = mat_w%matrix_struct%nrow_block
+      n_b_col = mat_w%matrix_struct%ncol_block
+      ! Number of processors
+      n_p_row = mat_w%matrix_struct%context%num_pe(1)
+      n_p_col = mat_w%matrix_struct%context%num_pe(2)
+      ! Position of the first processor
+      p_row_0 = mat_w%matrix_struct%first_p_pos(1)
+      p_col_0 = mat_w%matrix_struct%first_p_pos(2)
+      ! Position of the actual processor
+      p_row_i = mat_w%matrix_struct%context%mepos(1)
+      p_col_i = mat_w%matrix_struct%context%mepos(2)
+#endif
+
+      ! Initialization
+      NULLIFY (mat_u)
+      NULLIFY (mat_s)
+      NULLIFY (scf_cntrl)
+
+      ALLOCATE (et(n_ao))
+      ALLOCATE (ei(n_ao))
+      et(:) = 0.0_dp
+      ei(:) = 0.0_dp
+
+      ! Diagonalization of S
+      CALL get_qs_env(qs_env, matrix_s=mat_s)
+      CALL cp_fm_create(matrix=mat_u, matrix_struct=mat_w%matrix_struct, &
+                        name='OVERLAP MATRIX EIGENVECTORS')
+      CALL copy_dbcsr_to_fm(mat_s(1)%matrix, ec%tmatf)
+      CALL choose_eigv_solver(ec%tmatf, mat_w, et)
+      CALL cp_fm_to_fm(ec%tmatf, ec%tmatr)
+      ei(:) = et(:)
+
+      ! Eigenvector quenching
+      CALL get_qs_env(qs_env, scf_control=scf_cntrl)
+
+      n_deps = 0
+      DO i = 1, n_ao
+         IF (et(i) < scf_cntrl%eps_eigval) THEN
+#if defined(__SCALAPACK)
+            ic = indxg2p(i, n_b_col, p_col_i, p_col_0, n_p_col)
+            IF (p_col_i == ic) THEN
+               ic_loc = indxg2l(i, n_b_col, p_col_i, p_col_0, n_p_col)
+               DO j = 1, n_ao
+                  ir = indxg2p(j, n_b_row, p_row_i, p_row_0, n_p_row)
+                  IF (p_row_i == ir) THEN
+                     ir_loc = indxg2l(j, n_b_row, p_row_i, p_row_0, n_p_row)
+                     mat_w%local_data(ir_loc, ic_loc) = 0.0_dp
+                  END IF
+               END DO
+            END IF
+#else
+            mat_w%local_data(1:n_ao, i) = 0.0_dp
+#endif
+            n_deps = n_deps + 1
+         END IF
+      END DO
+
+      IF (n_deps /= 0) THEN
+         CALL cp_warn(__LOCATION__, &
+                      "Overlap matrix exhibits linear dependencies. At least some "// &
+                      "eigenvalues have been quenched.")
+      END IF
+
+      CALL cp_fm_to_fm(mat_w, mat_u)
+
+      ! Forward ransformations S^{-1/2}
+      IF (output_unit > 0) &
+         WRITE (output_unit, '(/,T3,A)') &
+         'Constructing forward basis-set transformation matrix'
+
+      DO i = 1, n_ao
+         IF (et(i) >= scf_cntrl%eps_eigval) THEN
+            f = et(i)**(-0.25_dp)
+#if defined(__SCALAPACK)
```

Comment:
Please use [cp_fm_set_all](https://github.com/cp2k/cp2k/blob/master/src/fm/cp_fm_types.F#L365) instead. Generally, high level code like this should not contain any `#if defined`.

Reply:
OK, I'll look at it, thanks!